### PR TITLE
[Table] Refactor RowHeader and ColumnHeader to compose with a shared component

### LINKED ISSUE #1109: [Table] Merge common ColumnHeader and RowHeader code into AbstractHeader
There's a lot of duplicate code now. Probably makes sense to introduce an `AbstractHeader` class that they both extend.

### PR DESCRIPTION
**Pull request by cmslewis**
####
- [x] Document how we should change the API in 2.0 now that I've combed through everything
- [x] Comb through shared-prop descriptions to make sure they're all generic (i.e., to make sure they don't mention only "row"s or "column"s).

#### Changes proposed in this pull request:

Goal here was to pull out the wealth of duplicated code between row-header and column-header code.

- Introduce `AbstractHeader`; have `RowHeader` and `ColumnHeader` compose with it.
- Introduce `AbstractHeaderCell`; have `RowHeaderCell` and `ColumnHeaderCell` compose with it.

#### Reviewers should focus on:

- ~This requires a bunch of abstract methods, which gets the job done but feels a little verbose. Would you prefer another approach?~ **Verdict: use composition instead**
- A lot of props between row-specific code and column-specific code have needlessly different names that make abstracting more convoluted (e.g. `isRowSelected` and `isColumnSelected`, instead of the more generic `isSelected`). In 2.0, we should genericize all of these props to make the APIs consistent. That'll simplify the code a bit.
- All tests still pass for me locally when running `gulp karma-table`; we'll see what happens here.

### CODE REVIEW

File: packages/table/src/headers/abstractHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,292 @@
+/**
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
```
Comment (cmslewis) [inline packages/table/src/headers/abstractHeader.tsx]: 2017

File: packages/table/src/headers/abstractHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    isResizable?: boolean;
+
+    /**
+     * The grid computes sizes of cells, rows, or columns from the
+     * configurable `columnWidths` and `rowHeights`.
+     */
+    grid: Grid;
+
+    /**
+     * If true, all `RowHeaderCell`s render their loading state except for those
```
Comment (cmslewis) [inline packages/table/src/headers/abstractHeader.tsx]: Genericize.

File: packages/table/src/headers/abstractHeader.tsx [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+    protected handleDragSelectableSelectionEnd() {
+        this.activationIndex = null; // not strictly required, but good practice
+        this.setState({ hasSelectionEnded: true });
+    }
+
+    protected isCellCurrentlyReorderable(isSelected: boolean) {
+        const { selectedRegions } = this.props;
+        // although reordering may be generally enabled for this row/column (via props.isReorderable), the
+        // row/col shouldn't actually become reorderable from a user perspective until a few other
```
Comment (cmslewis) [inline packages/table/src/headers/abstractHeader.tsx]: `col => `column`

File: packages/table/src/headers/abstractHeaderCell.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    isActive?: boolean;
+
+    /**
+     * The name displayed in the header of the row/column.
+     */
+    name?: string;
+
+    /**
+     * If `true`, the row `name` will be replaced with a fixed-height skeleton and the `resizeHandle`
```
Comment (cmslewis) [inline packages/table/src/headers/abstractHeaderCell.tsx]: Genericize.

File: packages/table/src/interactions/reorderable.tsx [hunk 5]
```diff
@@ -148,6 +148,7 @@ export class DragReorderable extends React.Component<IDragReorderable, {}> {
         this.props.onReordered(oldIndex, reorderedIndex, length);
 
         const newRegion = this.props.toRegion(reorderedIndex, reorderedIndex + length - 1);
+
```
Comment (cmslewis) [inline packages/table/src/interactions/reorderable.tsx]: Undo.

**Comment by adidahiya** [comment]
+1, use composition https://facebook.github.io/react/docs/composition-vs-inheritance.html

File: packages/table/src/headers/abstractHeader.tsx [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    headerCellIsSelectedPropName: string;
+
+    /**
+     * The maximum permitted size of the header in pixels. Corresponds to a width for column headers and
+     * a height for row headers.
+     */
+    maxSize: number;
+
+    /**
+     * The maximum permitted size of the header in pixels. Corresponds to a width for column headers and
```
Comment (cmslewis) [inline packages/table/src/headers/abstractHeader.tsx]: `minimum`

File: packages/table/src/headers/abstractHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+     */
+    resizeOrientation: Orientation;
+
+    /**
+     * The lowest cell index to render.
+     */
+    startIndex: number;
+
+    /**
+     * Converts a point on the screen to an index in the table grid.
```
Comment (cmslewis) [inline packages/table/src/headers/abstractHeader.tsx]: `to a row or column index`

File: packages/table/src/headers/abstractHeader.tsx [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+     */
+    isGhostIndex: (index: number) => boolean;
+
+    /**
+     * A callback that renders a ghost cell for the provided index.
+     */
+    renderGhostCell: (index: number, extremaClasses: string[]) => JSX.Element;
+
+    /**
+     * A callback that renders a non-ghost header cell at the provided index.
```
Comment (cmslewis) [inline packages/table/src/headers/abstractHeader.tsx]: `regular header cell`?

File: packages/table/src/headers/columnHeader.tsx [hunk 9]
```diff
@@ -43,102 +37,90 @@ export interface IColumnHeaderProps extends
...
+onSelection: (regions: IRegion[]) => void;
+selectedRegions: IRegion[];
+
+allowMultipleSelection: boolean;
+onFocus: (focusedCell: IFocusedCellCoordinates) => void;
+onSelection: (regions: IRegion[]) => void;
+onSelectionEnd?: (regions: IRegion[]) => void;
+selectedRegions: IRegion[];
+selectedRegionTransform?: ISelectedRegionTransform;
+*/
```
Comment (cmslewis) [inline packages/table/src/headers/columnHeader.tsx]: Delete

File: packages/table/src/table.tsx [hunk 10]
```diff
@@ -669,7 +669,6 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
             [Classes.TABLE_SELECTION_ENABLED]: this.isSelectionModeEnabled(RegionCardinality.FULL_COLUMNS),
         });
         const columnIndices = grid.getColumnIndicesInRect(viewportRect, fillBodyWithGhostCells);
-
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: Undo

**Comment by cmslewis** [comment]
@giladgray @adidahiya - now using composition. ready for another look.

File: packages/table/src/headers/abstractHeaderCell.tsx [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    shallowlyComparablePropKeys?: string[];
+}
+
+export interface IAbstractHeaderCellState {
+    isActive: boolean;
+}
+
+@ContextMenuTarget
+export abstract class AbstractHeaderCell  extends React.Component<IAbstractHeaderCellProps, IAbstractHeaderCellState> {
```
Comment (adidahiya) [inline packages/table/src/headers/abstractHeaderCell.tsx]: this doesn't need to be `abstract` anymore. can just be `export class HeaderCell` (same with the filename). in fact, it's kind of a TS bug that it allows you to write `<AbstractHeaderCell>` in JSX as-is right now.

File: packages/table/src/headers/abstractHeaderCell.tsx [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    public state = {
```
Comment (adidahiya) [inline packages/table/src/headers/abstractHeaderCell.tsx]: I don't think this gets type-checked unless you add the typedef `public state: IAbstractHeaderCellState`

File: packages/table/src/headers/columnHeader.tsx [hunk 13]
```diff
@@ -43,102 +37,67 @@ export interface IColumnHeaderProps extends
...
+                onResizeGuide={this.props.onResizeGuide}
+                onSelection={this.props.onSelection}
+                renderGhostCell={this.renderGhostCell}
+                renderHeaderCell={this.props.cellRenderer}
+                resizeOrientation={Orientation.VERTICAL}
+                selectedRegions={this.props.selectedRegions}
+                selectedRegionTransform={this.props.selectedRegionTransform}
+                startIndex={this.props.columnIndexStart}
+                toRegion={this.toRegion}
+                viewportRect={this.props.viewportRect}
```
Comment (giladgray) [inline packages/table/src/headers/columnHeader.tsx]: 😱  any sense in spreading `{...this.props}` to condense this _massive_ list?

File: packages/table/src/headers/abstractHeaderCell.tsx (hunk 11 above)
Comment (cmslewis) [inline packages/table/src/headers/abstractHeaderCell.tsx]: Fixed

File: packages/table/src/headers/abstractHeaderCell.tsx (hunk 12 above)
Comment (cmslewis) [inline packages/table/src/headers/abstractHeaderCell.tsx]: Fixed

File: packages/table/src/headers/abstractHeaderCell.tsx [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    shallowlyComparablePropKeys?: string[];
+}
+
+export interface IAbstractHeaderCellState {
+    isActive: boolean;
+}
+
+@ContextMenuTarget
+export class AbstractHeaderCell  extends React.Component<IAbstractHeaderCellProps, IAbstractHeaderCellState> {
```
Comment (giladgray) [inline packages/table/src/headers/abstractHeaderCell.tsx]: @adidahiya also called for this to be renamed to `HeaderCell` as it is no longer abstract. Please rename file accordingly. Same with other `abstract` things at this point...

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 15]
```diff
@@ -147,48 +87,42 @@ export class ColumnHeaderCell extends React.Component<IColumnHeaderCellProps, IC
...
-    };
-
-    public shouldComponentUpdate(nextProps: IColumnHeaderCellProps) {
-        // shallowly comparable props like "className" tend not to change in the default table
-        // implementation, so do that check last with hope that we return earlier and avoid it
-        // altogether.
-        return !Utils.shallowCompareKeys(this.props, nextProps, UPDATE_PROPS_KEYS)
-            || !Utils.deepCompareKeys(this.props.style, nextProps.style);
-    }
+    private static SHALLOWLY_COMPARABLE_PROP_KEYS = [
```
Comment (giladgray) [inline packages/table/src/headers/columnHeaderCell.tsx]: 👍  nice rename

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 16]
```diff
@@ -147,48 +87,42 @@ export class ColumnHeaderCell extends React.Component<IColumnHeaderCellProps, IC
...
+                className={this.props.className}
+                isActive={this.props.isActive}
+                isReorderable={this.props.isColumnReorderable}
+                isSelected={this.props.isColumnSelected}
+                loading={this.props.loading}
+                menu={this.props.menu}
+                name={this.props.name}
+                resizeHandle={this.props.resizeHandle}
+                shallowlyComparablePropKeys={ColumnHeaderCell.SHALLOWLY_COMPARABLE_PROP_KEYS}
+                style={this.props.style}
```
Comment (giladgray) [inline packages/table/src/headers/columnHeaderCell.tsx]: again, possible to spread `{...this.props}`?

File: packages/table/src/headers/abstractHeaderCell.tsx (hunk 14 above)
Comment (cmslewis) [inline packages/table/src/headers/abstractHeaderCell.tsx]: Not all props should be passed down (in particular, `isColumnReorderable` and `isColumnSelected`, whose names are specific to this component). However, can do something like:

```
const { isColumnReorderable, isColumnSelected, ...spreadableProps } = this.props;
return <AbstractHeader {...this.spreadableProps} ...etc...  />;
```

File: packages/table/src/headers/columnHeaderCell.tsx (hunk 16 above)
Comment (cmslewis) [inline packages/table/src/headers/columnHeaderCell.tsx]: Will play.

File: packages/table/src/headers/columnHeader.tsx (hunk 13 above)
Comment (cmslewis) [inline packages/table/src/headers/columnHeader.tsx]: Not all props should be passed down (in particular, isColumnReorderable and isColumnSelected, whose names are specific to this component). However, can do something like:

const { isColumnReorderable, isColumnSelected, ...spreadableProps } = this.props;
return <AbstractHeader {...this.spreadableProps} ...etc...  />;

File: packages/table/src/headers/abstractHeaderCell.tsx (hunk 14 above)
Comment (cmslewis) [inline packages/table/src/headers/abstractHeaderCell.tsx]: Done

**Comment by adidahiya** [comment]
`packages/table/src/headers/abstractHeader.tsx` seems to still be hanging around as an empty file

**Pull request merged**